apps/src/studio/Sprite.js

```diff
@@ -0,0 +1,459 @@
+  if (this.animation_) {
+    this.animation_.removeElement();
+  }
+  if (this.legacyAnimation_) {
+    this.legacyAnimation_.removeElement();
+  }
+};
+
+/**
+ * Retrieve animation frame duration (frames per tick)
+ */
+Sprite.prototype.getAnimationFrameDuration = function () {
+  if (this.dir === Direction.NONE) {
+    return constants.DEFAULT_SPRITE_ANIMATION_FRAME_DURATION;
+  } else {
+    return constants.DEFAULT_SPRITE_ANIMATION_FRAME_DURATION *
+        constants.DEFAULT_SPRITE_SPEED / this.speed;
+  }
+};
+
+/**
+ * Returns true if the item is currently fading away.
+ */
+Sprite.prototype.isFading = function() {
+  return !!this.startFadeTime;
+};
+
+/**
+ * Returns true if the item has finished fading away.  The caller will usually
+ * then call removeElement to destroy this item's assets.
+ */
+Sprite.prototype.hasCompletedFade = function() {
+  var currentTime = new Date().getTime();
+
+  return this.startFadeTime && currentTime > this.startFadeTime + this.fadeTime;
+};
+
+/**
+ * Display our item at its current location
+ */
+Sprite.prototype.display = function () {
+  var currentTime = new Date().getTime();
+  var opacity = 1;
+  if (this.startFadeTime) {
+    opacity = 1 - (currentTime - this.startFadeTime) / this.fadeTime;
+    opacity = Math.max(opacity, 0);
+    this.setOpacity(opacity);
+    if (this.hasCompletedFade()) {
+      // NOTE: we don't automatically change the state to hidden or set visible
+      // to false here.
+      this.startFadeTime = null;
+    }
+  }
+
+  var useLegacyAnimation = false;
+  var animationType;
+  var animationIndex;
+  var standingStill = this.displayDir === Direction.NONE;
+  var facingSouthWithEmotion =
+      this.displayDir === Direction.SOUTH && this.emotion !== Emotions.NORMAL;
+
+  if (standingStill || (!this.animation_ && facingSouthWithEmotion)) {
+    // Show idle animation while standing still
+    // if we only have a legacy animation, also show while moving south
+    animationIndex = this.emotion;
+    animationType = 'idle';
+
+    if (standingStill && this.frameCounts.normal) {
+      // If we see legacy normal frames (which are "idle" animations), use them:
+      useLegacyAnimation = true;
+    } else if (this.animation_ && !this.frameCounts.idleNormal) {
+      // If we are playing an "idle" animation from the primary spritesheet and
+      // there were no "normal" idle animations in the sheet, index based on
+      // (emotion - 1) instead of (emotion)
+      animationIndex -= 1;
+    }
+  } else if (facingSouthWithEmotion &&
+      this.animation_ && this.animation_.hasType('walkingEmotions')) {
+    animationIndex = this.emotion - 1;
+    animationType = 'walkingEmotions';
+  } else {
+    animationIndex = this.getDirectionFrame();
+    animationType = 'direction';
+  }
+  if (!this.animation_) {
+    useLegacyAnimation = true;
+  }
+
+  var drawPosition = this.getCurrentDrawPosition();
+
+  if (useLegacyAnimation) {
+    // Legacy render path:
```

Comment:
would it make sense to have specialised subclasses for non-legacy vs. legacy?  that way we could put the legacy codepath in a pretty explicit escrow by not touching its source file any more.